src/Language/PureScript/TypeChecker/Subsumption.hs

```diff
@@ -103,25 +99,23 @@ subsumes' SElaborate (ConstrainedType _ con ty1) ty2 = do
   let addDicts val = App val (TypeClassDictionary con dicts hints)
   return (elaborate . addDicts)
 subsumes' mode (TypeApp _ f1 r1) (TypeApp _ f2 r2) | eqType f1 tyRecord && eqType f2 tyRecord = do
-    let goWithLabel l t1 t2 = withErrorMessageHint (ErrorInRowLabel l) $ subsumes' SNoElaborate t1 t2
-    let (common, ((ts1', r1'), (ts2', r2'))) = alignRowsWith goWithLabel r1 r2
-    -- For { ts1 | r1 } to subsume { ts2 | r2 } when r1 is empty (= we're working with a closed row),
-    -- every property in ts2 must appear in ts1. If not, then the candidate expression is missing a required property.
-    -- Conversely, when r2 is empty, every property in ts1 must appear in ts2, or else the expression has
-    -- an additional property which is not allowed.
-    when (isREmpty r1')
-      (for_ (firstMissingProp ts2' ts1') (throwError . errorMessage . PropertyIsMissing . rowListLabel))
-    when (isREmpty r2')
-      (for_ (firstMissingProp ts1' ts2') (throwError . errorMessage . AdditionalProperty . rowListLabel))
+  let 
+    withBoolError :: forall a. m a -> m Bool
+    withBoolError u = catchError (True <$ u) (pure . const False)
+  unifyishRows
     -- Check subsumption for common labels
-    sequence_ common
+    (subsumes' SNoElaborate) 
     -- Inject the info here
-    unifyTypes (rowFromList (ts1', r1')) (rowFromList (ts2', r2'))
-    -- Nothing was elaborated, return the default coercion
-    return (defaultCoercion mode)
-  where
-    -- Find the first property that's in the first list (of tuples) but not in the second
-    firstMissingProp t1 t2 = fst <$> uncons (minusBy' (comparing rowListLabel) t1 t2)
+    (uncurry $ \(ts1', r1') (ts2', r2') ->
+      withBoolError $ unifyTypes (rowFromList (ts1', r1')) (rowFromList (ts2', r2')))
+    isTypesDoNotUnify
+    (\r1' r2' -> do
+      subst <- gets checkSubstitution
+      pure $ errorMessage $ TypesDoNotUnify (substituteType subst r1') (substituteType subst r2'))
```

Comment:
All of this is shared with the true `unifyTypes`, right? (Or at least, with access to `unifyTails` instead of the `withBoolError` hack, it would be.) How about exporting another variation from TypeChecker.Types for that case and subsumption?

I don't know what to call it, though. `unifyishTypes` is already hard to take seriously. :laughing: